### DISCUSSION TITLE
Ideas for future funding opportunities

### DISCUSSION THREAD

**Discussion by robgjansen** [shadow/shadow#2995] [Ideas]
Below is the start of a list of rough ideas that could go into future proposals to obtain funding for Shadow development.
    
- Distributed execution (multiple machines)
- Modeling/simulating CPU execution time
- Fuzzing / fault injection framework for exposing bugs in managed processes
- Virtual filesystem (to enforce determinism on FS operations)
- TUI
- Break out pieces into generally usable Rust crates and publish them
  - Utilities for working with shared memory (I haven't found any such things elsewhere):
      - shmem allocator
      - VASI trait and macro
      - vasi-sync
  - no_std + no-libc crates for kernel APIs (WIP)
  - Maybe a more general framework for intercepting + emulating syscalls:
      - split core functionality into a crate that Shadow uses?
      - add hooks into Shadow for injecting custom syscall handlers?
      - both?
  - VDSO dynamic hijacking (e.g. for intercepting time VDSO fns)
- NETLINK subsystem (esp. if we have compelling applications that want it)
  - https://man7.org/linux/man-pages/man7/netlink.7.html
- CMSG subsystem (esp. if we have compelling applications that want it)
  - https://linux.die.net/man/3/cmsg
  - CSMG can be used in e.g. [`recvmsg`](https://linux.die.net/man/2/recvmsg)

**Comment by jtracey** [comment]
On the note of "Break out pieces into generally usable Rust crates and publish them", a (possibly ambitious) project I recently wished existed was the ability to use Shadow as a Rust library for tests. This would allow, e.g., running hours worth of light network unit tests or fuzzing in a few seconds. For my project, the workaround was of course to use standalone integration tests with Shadow, but it would have been nice to have tightly integrated Rust functionality so you don't have to use an entire project to test a particular code path, or need a standalone Shadow install.

**Reply by stevenengler** [reply]
In your example do you mean shadow would be enabled on-demand within the current process (shadow's interposition toggled on when the test starts, and then off once the test is finished), or would it be something that would fork the test and run the test in a different process? Or would this distinction matter for your use case?

**Reply by jtracey** [reply]
I don't think it would matter for my use case, mostly I just had in mind something to test particular functions in Shadow when running `cargo test`; I don't have strong design opinions, and haven't spent much time thinking about how feasible it'd be.

**Comment by stevenengler** [comment]
There are also some ideas in the ["Limitations and workarounds"](https://shadow.github.io/docs/guide/limitations.html) docs:

- IPv6
- statically linked executables
